File: jane_street/market_prediction/cpp/src/engine.cpp
#include "market_prediction/engine.hpp"

namespace market_prediction {

InferenceEngine::InferenceEngine(std::shared_ptr<ModelManager> model_manager, 
                                 std::shared_ptr<MarketPreprocessor> preprocessor,
                                 double threshold)
    : model_manager_(model_manager), preprocessor_(preprocessor), threshold_(threshold) {}

int InferenceEngine::predict_action(const Eigen::VectorXd& features) {
    Eigen::VectorXd processed = preprocessor_->transform(features);
    Eigen::VectorXd prob = model_manager_->predict(processed);
    return prob(0) >= threshold_ ? 1 : 0;
}

std::vector<int> InferenceEngine::batch_predict(const Eigen::MatrixXd& features_batch) {
    std::vector<int> actions;
    actions.reserve(features_batch.rows());
    
    // Vectorized preprocessor transform for whole batch would be better, 
    // but here we maintain same logic as python version.
    for (int i = 0; i < features_batch.rows(); ++i) {
        actions.push_back(predict_action(features_batch.row(i)));
    }
    
    return actions;
}

} // namespace market_prediction

File: jane_street/market_prediction/rust/src/engine.rs
use crate::models::ModelManager;
use crate::preprocessor::MarketPreprocessor;
use ndarray::Array1;

pub struct InferenceEngine {
    model_manager: ModelManager,
    preprocessor: MarketPreprocessor,
    threshold: f64,
}

impl InferenceEngine {
    pub fn new(model_manager: ModelManager, preprocessor: MarketPreprocessor, threshold: f64) -> Self {
        Self {
            model_manager,
            preprocessor,
            threshold,
        }
    }

    pub fn predict_action(&mut self, features: &Array1<f64>) -> i32 {
        let processed = self.preprocessor.transform(features);
        let prob = self.model_manager.predict(&processed);
        if prob >= self.threshold { 1 } else { 0 }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_engine_predict() {
        let model_manager = ModelManager::new(136);
        let preprocessor = MarketPreprocessor::new(130, 5);
        let mut engine = InferenceEngine::new(model_manager, preprocessor, 0.5);
        
        let features = Array1::from_elem(130, 0.5);
        let action = engine.predict_action(&features);
        assert!(action == 0 || action == 1);
    }
}
